lib/ansible/modules/windows/win_psmodule.ps1

```diff
@@ -1,186 +1,380 @@
+                $ht.Add("AllVersions", $true)
+            }
+
+            if ( $ExistingModuleBefore.Exists) {
+                # The Force parameter overwrite the WhatIf parameter
+                if ( -not $CheckMode ) {
+                    Uninstall-Module @ht -ErrorVariable ErrorDetails | out-null
+                }
+                $result.changed = $true
+            }
+        }
+        catch{
+            $ErrorMessage = "Problems removing $($Name) module: $($ErrorDetails.Exception.Message)"
+            Fail-Json $result $ErrorMessage
+        }
     }
-    else {      
-      try{
-        # Install NuGet Provider if needed
-        Install-NugetProvider -CheckMode $CheckMode;
+}
 
+Function Find-LatestPsModule {
+    Param(
+        [Parameter(Mandatory=$true)]
+        [String]$Name,
+        [String]$Repository,
+        [Bool]$AllowPrerelease,
+        [Bool]$CheckMode
+    )
+
+    try {
         $ht = @{
-            Name      = $Name;
-            WhatIf    = $CheckMode;
-            ErrorAction = "Stop";
-            Force     = $true;
-        };
-
-        # If specified, use repository name to select module source
-        if ($Repository) {
-            $ht["Repository"] = "$Repository";
-        }
-
-        # Check Powershell Version (-AllowClobber was introduced in PowerShellGet 1.6.0)
-        if ("AllowClobber" -in ((Get-Command PowerShellGet\Install-Module | Select -ExpandProperty Parameters).Keys)) {
-          $ht['AllowClobber'] = $AllowClobber;
-        }
-        
-        Install-Module @ht | out-null;
-        
-        $result.output = "Module $($Name) installed"
-        $result.changed = $true
-      }
-      catch{
-        $ErrorMessage = "Problems installing $($Name) module: $($_.Exception.Message)"
+            Name = $Name
+        }
+
+        if ( $AllowPrerelease ) {
+            $ht.Add("AllowPrerelease",$AllowPrerelease)
+        }
+
+        # If specified, use repository name to select module source.
+        if ( $Repository ) {
+            $ht.Add("Repository", "$Repository")
+        }
+
+        $LatestModule = Find-Module @ht
+        $LatestModuleVersion = $LatestModule.Version
+    }
+    catch {
+        $ErrorMessage = "Cant find the module $($Name): $($_.Exception.Message)"
         Fail-Json $result $ErrorMessage
-      }
     }
+
+    $LatestModuleVersion
 }
 
-Function Remove-PsModule {
-    param(
-      [Parameter(Mandatory=$true)]
-      [string]$Name,
-      [bool]$CheckMode
-    )
-    # If module is present, unistalls it.
-    if (Get-Module -Listavailable|?{$_.name -eq $Name}){
-      try{
-        Uninstall-Module -Name $Name -Confirm:$false -Force -ErrorAction Stop -WhatIf:$CheckMode | out-null
-        $result.output = "Module $($Name) removed"
-        $result.changed = $true
-      }
-      catch{
-        $ErrorMessage = "Problems removing $($Name) module: $($_.Exception.Message)"
+# Check PowerShell version, fail if < 5.0
+$PsVersion = $PSVersionTable.PSVersion
```

Comment:
We should remove this check as technically older PowerShell versions can have PowerShellGet installed. Best to change this to check if PowerShellGet is installed with the required cmdlets.